fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/prunesubfield/SubfieldAccessPathNormalizer.java

```diff
@@ -98,84 +103,84 @@ public boolean hasPath(ColumnRefOperator root) {
+            if (!PruneSubfieldRule.SUPPORT_FUNCTIONS.contains(call.getFnName())) {
+                return Optional.empty();
+            }
 
-        if (!collectionElementOp.getChild(1).isConstant()) {
-            currentPath.appendPath(ColumnAccessPath.PATH_PLACEHOLDER, TAccessPathType.ALL);
-            return null;
-        }
-        currentPath.appendPath(ColumnAccessPath.PATH_PLACEHOLDER, TAccessPathType.INDEX);
-        return null;
-    }
+            if (call.getFnName().equals(FunctionSet.MAP_KEYS)) {
+                return childrenAccessPaths.get(0)
+                        .map(p -> p.appendPath(ColumnAccessPath.PATH_PLACEHOLDER, TAccessPathType.KEY));
+            } else if (FunctionSet.MAP_SIZE.equals(call.getFnName())
+                    || FunctionSet.CARDINALITY.equals(call.getFnName())
+                    || FunctionSet.ARRAY_LENGTH.equals(call.getFnName())) {
+                return childrenAccessPaths.get(0)
+                        .map(p -> p.appendPath(ColumnAccessPath.PATH_PLACEHOLDER, TAccessPathType.OFFSET));
+            }
 
-    @Override
-    public Void visitCall(CallOperator call, Void context) {
-        if (!PruneSubfieldRule.SUPPORT_FUNCTIONS.contains(call.getFnName())) {
-            return visit(call, context);
+            return Optional.empty();
         }
 
-        if (call.getFnName().equals(FunctionSet.MAP_KEYS)) {
-            call.getChild(0).accept(this, context);
-            if (currentPath != null) {
-                currentPath.appendPath(ColumnAccessPath.PATH_PLACEHOLDER, TAccessPathType.KEY);
-            }
-            return null;
-        } else if (FunctionSet.MAP_SIZE.equals(call.getFnName())
-                || FunctionSet.CARDINALITY.equals(call.getFnName())
-                || FunctionSet.ARRAY_LENGTH.equals(call.getFnName())) {
-            call.getChild(0).accept(this, context);
-            if (currentPath != null) {
-                currentPath.appendPath(ColumnAccessPath.PATH_PLACEHOLDER, TAccessPathType.OFFSET);
+        private Optional<AccessPath> process(ScalarOperator scalarOperator, Deque<AccessPath> accessPaths) {
+            // process children in post-order
+            List<Optional<AccessPath>> childAccessPaths = scalarOperator.getChildren().stream()
+                    .map(child -> process(child, accessPaths))
+                    .collect(Collectors.toList());
+            // no AccessPaths gathered from children of intermediate ScalarOperator means current
+            // scalar operator contains not nested types.
+            if (!childAccessPaths.isEmpty() && childAccessPaths.stream().noneMatch(Optional::isPresent)) {
+                return Optional.empty();
             }
+            Optional<AccessPath> currentPath = scalarOperator.accept(this, childAccessPaths);
+            AccessPath path = currentPath.orElse(null);
+            childAccessPaths.stream().filter(p -> p.isPresent() && p.get() != path)
```

Comment:
```suggestion
            childAccessPaths.stream().distinct()
```
and add some comment?